osquery/events/eventfactory.cpp

```diff
@@ -0,0 +1,478 @@
+        if (item.GetString() == name) {
+          VLOG(1) << "Disabling event subscriber: " << name;
+          specialized_sub->disabled = true;
+        }
+      }
+    }
+  }
+
+  if (specialized_sub->state() != EventState::EVENT_NONE) {
+    specialized_sub->tearDown();
+  }
+
+  // Allow subscribers a configure-time setup to determine if they should run.
+  auto status = specialized_sub->setUp();
+  if (!status) {
+    specialized_sub->disabled = true;
+  }
+  specialized_sub->state(EventState::EVENT_SETUP);
+
+  // Let the subscriber initialize any Subscriptions.
+  if (!FLAGS_disable_events && !specialized_sub->disabled) {
+    status = specialized_sub->init();
+    specialized_sub->state(EventState::EVENT_RUNNING);
+  } else {
+    specialized_sub->state(EventState::EVENT_PAUSED);
+  }
+
+  auto& ef = EventFactory::getInstance();
+  {
+    RecursiveLock lock(ef.factory_lock_);
+    ef.event_subs_[name] = specialized_sub;
+  }
+
+  // Set state of subscriber.
+  if (!status.ok()) {
+    specialized_sub->state(EventState::EVENT_FAILED);
+    return Status(1, status.getMessage());
+  } else {
+    return Status::success();
+  }
+}
+
+Status EventFactory::registerEventPublisher(const PluginRef& pub) {
+  // Try to downcast the plugin to an event publisher.
+  EventPublisherRef specialized_pub;
+  try {
+    auto base_pub = std::dynamic_pointer_cast<EventPublisherPlugin>(pub);
+    specialized_pub = std::static_pointer_cast<BaseEventPublisher>(base_pub);
+  } catch (const std::bad_cast& /* e */) {
+    return Status(1, "Incorrect plugin");
+  }
+
+  if (specialized_pub == nullptr || specialized_pub.get() == nullptr) {
+    return Status::success();
+  }
+
+  auto type_id = specialized_pub->type();
+  if (type_id.empty()) {
+    // This subscriber did not override its name.
+    return Status(1, "Publishers must have a type");
+  }
+
+  auto& ef = EventFactory::getInstance();
+  {
+    RecursiveLock lock(ef.factory_lock_);
+    if (ef.event_pubs_.count(type_id) != 0) {
+      // This is a duplicate event publisher.
+      return Status(1, "Duplicate publisher type");
+    }
+
+    ef.event_pubs_[type_id] = specialized_pub;
+  }
+
+  // Do not set up event publisher if events are disabled.
+  if (!FLAGS_disable_events) {
+    if (specialized_pub->state() != EventState::EVENT_NONE) {
+      specialized_pub->tearDown();
+    }
+
+    auto status = specialized_pub->setUp();
+    specialized_pub->state(EventState::EVENT_SETUP);
+    if (!status.ok()) {
+      // Only start event loop if setUp succeeds.
+      LOG(INFO) << "Event publisher not enabled: " << type_id << ": "
+                << status.what();
+      specialized_pub->isEnding(true);
+      return status;
+    }
+  }
+
+  return Status::success();
+}
+
+Status EventFactory::deregisterEventSubscriber(const std::string& sub) {
+  auto& ef = EventFactory::getInstance();
+
+  RecursiveLock lock(ef.factory_lock_);
+  if (ef.event_subs_.count(sub) == 0) {
```

Comment:
heads up there's a `.count` then `.at` here that can be replaced with `.find`

Reply:
see see https://github.com/osquery/osquery/pull/6610#discussion_r540330119